aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -118,15 +118,49 @@ impl<'e, E: ExecutorView> StorageAdapter<'e, E> {
             };
 
             let buf_size = resource_size(&buf);
-            Ok((buf, buf_size + group_size as usize))
-        } else {
-            let state_key = resource_state_key(address, struct_tag)?;
-            let buf = self
-                .executor_view
-                .get_resource_bytes(&state_key, maybe_layout)?;
-            let buf_size = resource_size(&buf);
-            Ok((buf, buf_size))
+            return Ok(Some((buf, buf_size + group_size as usize)));
         }
+        Ok(None)
+    }
+
+    fn get_any_resource_with_layout(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+        metadata: &[Metadata],
+        maybe_layout: Option<&MoveTypeLayout>,
+    ) -> PartialVMResult<(Option<Bytes>, usize)> {
+        let resource_group_bytes =
+            self.get_resource_group_bytes(address, struct_tag, metadata, maybe_layout)?;
+        if let Some(result) = resource_group_bytes {
+            return Ok(result);
+        }
+
+        let state_key = resource_state_key(address, struct_tag)?;
+        let buf = self
+            .executor_view
+            .get_resource_bytes(&state_key, maybe_layout)?;
+        let buf_size = resource_size(&buf);
+        Ok((buf, buf_size))
+    }
+
+    fn get_any_resource_size_with_layout(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+        metadata: &[Metadata],
+        maybe_layout: Option<&MoveTypeLayout>,
+    ) -> PartialVMResult<(Option<u64>, usize)> {
+        let resource_group_bytes =
+            self.get_resource_group_bytes(address, struct_tag, metadata, maybe_layout)?;
+        if let Some((bytes, bytes_loaded)) = resource_group_bytes {
+            return Ok((bytes.map(|bytes| bytes.len() as u64), bytes_loaded));
+        }
+        let state_key = resource_state_key(address, struct_tag)?;
+        let size = self
+            .executor_view
+            .get_resource_state_value_size(&state_key)?;
+        Ok((size, size.unwrap_or(0) as usize))
```

Comment:
nit: can we change resolver to use u64 as well? We cast to usize here to cast it to u64 back when we create `NumBytes` later. Or even we can return `NumBytes` directly, to indicate this is for gas charging